File: internal/indexer/extractors/golang.go
package extractors

import (
	"context"

	sitter "github.com/smacker/go-tree-sitter"
	"github.com/smacker/go-tree-sitter/golang"
)

type GoExtractor struct{}

func NewGoExtractor() *GoExtractor {
	return &GoExtractor{}
}

func (e *GoExtractor) Parse(content []byte) (*sitter.Tree, error) {
	parser := sitter.NewParser()
	parser.SetLanguage(golang.GetLanguage())
	return parser.ParseCtx(context.Background(), nil, content)
}

func (e *GoExtractor) Extract(tree *sitter.Tree, content []byte) ([]Entity, []Relation) {
	var entities []Entity
	var relations []Relation

	// Query for Go entities and relations
	q, err := sitter.NewQuery([]byte(`
		(function_declaration name: (identifier) @func.name) @func.def
		(method_declaration name: (field_identifier) @method.name) @method.def
		(type_spec name: (type_identifier) @type.name) @type.def
		(call_expression function: (identifier) @call.name) @call.def
	`), golang.GetLanguage())
	if err != nil {
		return entities, relations
	}

	qc := sitter.NewQueryCursor()
	qc.Exec(q, tree.RootNode())

	type callInfo struct {
		Name string
		Line int
	}
	var calls []callInfo

	for {
		m, ok := qc.NextMatch()
		if !ok {
			break
		}
		
		qName := ""
		nodeType := ""
		start := 0
		end := 0

		for _, c := range m.Captures {
			name := q.CaptureNameForId(c.Index)
			if name == "func.name" || name == "method.name" || name == "type.name" || name == "call.name" {
				qName = c.Node.Content(content)
			}
			if name == "func.def" {
				nodeType = TypeFunction
				start = int(c.Node.StartPoint().Row) + 1
				end = int(c.Node.EndPoint().Row) + 1
			} else if name == "method.def" {
				nodeType = TypeMethod
				start = int(c.Node.StartPoint().Row) + 1
				end = int(c.Node.EndPoint().Row) + 1
			} else if name == "type.def" {
				nodeType = TypeType
				start = int(c.Node.StartPoint().Row) + 1
				end = int(c.Node.EndPoint().Row) + 1
			} else if name == "call.def" {
				nodeType = "CALL"
				start = int(c.Node.StartPoint().Row) + 1
			}
		}

		if qName != "" && nodeType != "" {
			if nodeType == "CALL" {
				calls = append(calls, callInfo{Name: qName, Line: start})
			} else {
				entities = append(entities, Entity{
					Name:          qName,
					QualifiedName: qName, // Needs full resolution in advanced parsers
					Type:          nodeType,
					LineStart:     start,
					LineEnd:       end,
				})
			}
		}
	}

	// Heuristic relation mapping
	for _, call := range calls {
		for _, ent := range entities {
			if call.Line >= ent.LineStart && call.Line <= ent.LineEnd {
				relations = append(relations, Relation{
					Source: QualifiedName(ent.QualifiedName),
					Target: QualifiedName(call.Name),
					Type:   RelCalls,
				})
				break
			}
		}
	}

	return entities, relations
}

File: src/indexer/extractors/golang.rs
use super::base::{Entity, Extractor, Relation};
use anyhow::Result;
use tree_sitter::{Parser, Query, QueryCursor};

pub struct GoExtractor {
    parser: Parser,
    content: String,
    entities: Vec<Entity>,
    relations: Vec<Relation>,
}

impl GoExtractor {
    pub fn new() -> Result<Self> {
        let mut parser = Parser::new();
        // Since tree_sitter versions changed slightly, we rely on tree-sitter-go
        parser.set_language(&tree_sitter_go::language())?;
        Ok(Self {
            parser,
            content: String::new(),
            entities: Vec::new(),
            relations: Vec::new(),
        })
    }
}

impl Extractor for GoExtractor {
    fn parse(&mut self, content: &str) -> Result<()> {
        self.content = content.to_string();

        let tree = self
            .parser
            .parse(content, None)
            .ok_or_else(|| anyhow::anyhow!("Failed to parse Go code"))?;

        let language = tree_sitter_go::language();

        // 1. Query for Entities
        let query_str = "
            (function_declaration name: (identifier) @func body: (block) @body)
            (type_spec name: (type_identifier) @type)
            (method_declaration name: (field_identifier) @method body: (block)? @body)
        ";
        let query = Query::new(&language, query_str)?;
        let mut cursor = QueryCursor::new();

        let call_query = Query::new(&language, "(call_expression function: (identifier) @call)")?;

        let bindings = cursor.matches(&query, tree.root_node(), content.as_bytes());
        for m in bindings {
            let mut ent_name = String::new();
            let mut ent_type = "UNKNOWN";
            let mut body_node = None;
            let mut start_b = 0;
            let mut end_b = 0;

            for capture in m.captures {
                let name = capture.node.utf8_text(content.as_bytes())?.to_string();
                let capture_name = query.capture_names()[capture.index as usize];

                if capture_name == "func" || capture_name == "type" || capture_name == "method" {
                    ent_name = name;
                    ent_type = match capture_name {
                        "func" => "FUNCTION",
                        "type" => "TYPE",
                        "method" => "METHOD",
                        _ => "UNKNOWN",
                    };
                    let parent = capture.node.parent().unwrap_or(capture.node);
                    start_b = parent.start_byte();
                    end_b = parent.end_byte();
                } else if capture_name == "body" {
                    body_node = Some(capture.node);
                }
            }

            if !ent_name.is_empty() {
                self.entities.push(Entity {
                    name: ent_name.clone(),
                    entity_type: ent_type.to_string(),
                    qualified_name: ent_name.clone(),
                    start_byte: start_b,
                    end_byte: end_b,
                });

                // 2. Extract Relations (CALLS) if there's a body
                if let Some(body) = body_node {
                    let mut call_cursor = QueryCursor::new();
                    let call_bindings = call_cursor.matches(&call_query, body, content.as_bytes());
                    for cm in call_bindings {
                        for c in cm.captures {
                            let target_call = c.node.utf8_text(content.as_bytes())?.to_string();
                            self.relations.push(Relation {
                                source: ent_name.clone(),
                                target: target_call,
                                relation_type: "CALLS".to_string(),
                            });
                        }
                    }
                }
            }
        }

        Ok(())
    }

    fn extract(&self) -> Result<(Vec<Entity>, Vec<Relation>)> {
        Ok((self.entities.clone(), self.relations.clone()))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_go_extractor() {
        let code = r#"
package main

type Server struct {}

func (s *Server) Start() {
    InitDB()
    LogMessage()
}

func InitDB() {
}
        "#;

        let mut ext = GoExtractor::new().expect("Failed to init go extractor");
        ext.parse(code).expect("Failed to parse go code");

        let (entities, relations) = ext.extract().unwrap();

        let ent_names: Vec<&str> = entities.iter().map(|e| e.name.as_str()).collect();
        assert!(ent_names.contains(&"Server"));
        assert!(ent_names.contains(&"Start"));
        assert!(ent_names.contains(&"InitDB"));

        // Ensure CALLS relations exist
        let call_init = relations
            .iter()
            .find(|r| r.source == "Start" && r.target == "InitDB");
        let call_log = relations
            .iter()
            .find(|r| r.source == "Start" && r.target == "LogMessage");

        assert!(
            call_init.is_some(),
            "Relation Start -> CALLS -> InitDB missing"
        );
        assert!(
            call_log.is_some(),
            "Relation Start -> CALLS -> LogMessage missing"
        );
    }
}
